### ISSUE TITLE
Mono misinterpreting Corflags (?)

### ISSUE THREAD

**Issue by leezer3** [mono/mono#21038]
I first raised this as a bug with Mono.Cecil:
https://github.com/jbevain/cecil/issues/753

They suggest it's likely a Mono issue, and so I'm raising it here. Unfortunately, I'm a little out of my depth WRT to PE headers / corflags, but happy to play around & help in any way I can.

## Steps to Reproduce

1.  Compile exectuable as AnyCPU using Mono.
2. Use Mono.Cecil to modify the corflags value to 32-bit required.
3. Observe that under Linux (only), the modified exectuable is actually running under 64-bit.
4. Using the identical tool / compiled executable (copied to Windows machine / .Net machine), the modified 32-bit executable works as expected under Linux.

Relevant tool code (nothing special?)
```
ModuleDefinition modDef = ModuleDefinition.ReadModule(fileToProcess);
if (modDef == null)
{
	Console.WriteLine("An unexpected error occured whilst attempting to apply Corflags to executable " + fileToProcess);
	return;
}
modDef.Attributes |= ModuleAttributes.Required32Bit;
try
{
	modDef.Write(finalFileName);
}
catch
{
	Console.WriteLine("Failed to write 32-bit executable to " + finalFileName);
	throw;
}
```
https://github.com/leezer3/OpenBVE/blob/master/source/DevTools/LBAHeader/Program.cs#L68

### Current Behavior

Modified exectuable unexpected executes as 64-bit.

### Expected Behavior

Modified exectable exectutes as 32-bit.

## On which platforms did you notice this

[ ] macOS
[X] Linux
[ ] Windows

**Version Used**:

Mono 6.12.0.122 (Azure pipelines CI, don't have the full version stack off the top of my head)

Can also reproduce locally:
```
Mono JIT compiler version 6.12.0.90 (tarball Fri Sep  4 14:03:37 UTC 2020)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       altstack
        Notifications: epoll
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug
        Interpreter:   yes
        LLVM:          yes(610)
        Suspend:       hybrid
        GC:            sgen (concurrent by default)
```

## Stacktrace

n/a

**Comment by lambdageek** [comment]
Mono does not use the flags to decide how to run the app.

You can use `mono --arch=32` to start `mono32` (assuming you have it installed on your PATH) in 32-bit mode.

**Comment by leezer3** [comment]
OK, so that's interesting, thanks.

However, that option doesn't appear to be available / functional under Linux unless I'm missing something.
```
osboxes@osboxes:~/test/testapp$ mono test.exe
64-bit
osboxes@osboxes:~/test2/testapp$ mono test.exe --arch=32
64-bit
osboxes@osboxes:~/test2/WinModified$ mono --arch=32 test.exe
Unknown command line option: '--arch=32'
```

Dead simple test app this time:
```
using System;

namespace TestProgram
{
	class Program
	{
		static void Main(string[] args)
		{
			if (IntPtr.Size == 4)
			{
				Console.WriteLine("32-bit");
			}
			else
			{
				Console.WriteLine("64-bit");
			}
		}
	}
}
```

**Comment by lambdageek** [comment]
> OK, so that's interesting, thanks.
> 
> However, that option doesn't appear to be available / functional under Linux unless I'm missing something.
> 
> ```shell
> osboxes@osboxes:~/test2/WinModified$ mono --arch=32 test.exe
> Unknown command line option: '--arch=32'
> ```

Ah you're right.  Looks like that's only enabled on OSX.

You can install the i386 mono packages (although not at the same time as the amd64)

On Ubuntu 18.04:
```
$ dpkg --add-architecture i386
$ sudo apt install gnupg ca-certificates
$ sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 3FA7E0328081BFF6A14DA29AA6A19B38D3D831EF
$ echo "deb [arch=i386] https://download.mono-project.com/repo/ubuntu stable-bionic main" | sudo tee /etc/apt/sources.list.d/mono-official-stable-i386.list
$sudo apt update
```
(modifying the instructions from https://www.mono-project.com/download/stable/ slightly).

You should then be able to pick the i386 version of the various mono packages.

---

Is there some particular behavior you expect from a 32-bit runtime?

**Comment by leezer3** [comment]
Thanks, that explains a lot.
(It would be nice if there was a functional way to do this without the i386 packages, especially as x86 compiled executables work on the supplied Mono)

---

Expected behaviour-
Not as such, it's complicated, and as you might expect involves legacy code....

TLDR:
A mid-level component (legacy train plugins, of which we have a lot) involves direct DLL imports of Win32 DLL plugins via dllimport, either via Windows or WINE-mono.
Obviously, we can't do this in a pure 64-bit process, hence we've been stuck on 32-bit for quite a while, whilst we found a way around this.
The solution ended up being twofold- Creating a 32-bit proxy app (communicating with the main program via WCF), and marking a copy of the original executable as 32-bit preferered, which uses 32-bit on Win32 & allows the original in-process loading to be used if required.
Predictably, I promptly recieved a bug report that the 32-bit executable wasn't working on Linux, hence heading down this rabbit hole......
